src/devices/cddadevice.h

```diff
@@ -18,31 +18,42 @@
 #ifndef CDDADEVICE_H
 #define CDDADEVICE_H
 
-#include <QMutex>
+#include <QTimer>
 
 // These must come after Qt includes (issue 3247)
 #include <cdio/cdio.h>
-#include <gst/audio/gstaudiocdsrc.h>
 
 #include "cddasongloader.h"
 #include "connecteddevice.h"
 #include "core/song.h"
-#include "musicbrainz/musicbrainzclient.h"
+
+class QUrl;
 
 class CddaDevice : public ConnectedDevice {
   Q_OBJECT
 
  public:
+  // Initializes the CddaDevice but does not open a handle to the device.
+  // Using code MUST call Init() after construction before calling and
+  // check that it returns true any other methods, to ensure that a valid
+  // device handle was correctly opened. Using class methods without
+  // a valid device handle is undefined behavior and might result in crashes.
   Q_INVOKABLE CddaDevice(const QUrl& url, DeviceLister* lister,
                          const QString& unique_id, DeviceManager* manager,
                          Application* app, int database_id, bool first_time);
   ~CddaDevice();
 
-  void Init();
-  void Refresh();
+  bool Init() override;
   bool CopyToStorage(const MusicStorage::CopyJob&) { return false; }
   bool DeleteFromStorage(const MusicStorage::DeleteJob&) { return false; }
-
+  CddaSongLoader* loader();
+  // Access to the raw cdio device handle.
+  CdIo_t* raw_cdio();  // TODO: not ideal, but Ripper needs this currently
+  // Check whether a valid device handle was opened.
+  bool IsValid() const;
+  void WatchForDiscChanges(bool watch);
+
+  static const int kDiscChangePollingIntervalMs;
```

Comment:
I think this can be omitted from the header and be in an anonymous namespace in the `.cpp`

Reply:
I think `IsValid`  should be a public member as there is no other way to know whether any given instance of `CddaDevice` is properly set up.

`WatchForDiscChanges` modifies the instance state and it makes sense to me to keep it as a class method.

Reply:
I meant specifically `kDiscChangePollingIntervalMs` which I don't think needs to be here.

Reply:
I see. That's done.